File: src/backend/studio-reservation/src/main/java/org/taonaw/studio_reservation/domain/model/cancellationFeeSetting/CancellationFeeRate.java
package org.taonaw.studio_reservation.domain.model.cancellationFeeSetting;

import lombok.EqualsAndHashCode;
import lombok.Getter;
import lombok.NonNull;
import org.taonaw.studio_reservation.domain.shared.Assertion;

import java.time.LocalDateTime;

@Getter
@EqualsAndHashCode
public class CancellationFeeRate {
    private final int daysAgo;
    private final double rate;

    public CancellationFeeRate(int daysAgo, double rate) {
        Assertion.argumentRange(daysAgo, 0, 50);
        Assertion.argumentRange(rate, 0, 100);
        this.daysAgo = daysAgo;
        this.rate = rate;
    }

    public boolean isApplied(@NonNull LocalDateTime targetDateTime, @NonNull LocalDateTime currentDateTime) {
        var appliedDate = targetDateTime.toLocalDate().minusDays(daysAgo);
        var currentDate = currentDateTime.toLocalDate();
        return currentDate.isEqual(appliedDate) || currentDate.isAfter(appliedDate);
    }

    public boolean isFree() {
        return rate == 0;
    }

    public boolean isDuplicated(@NonNull CancellationFeeRate other) {
        return this != other && this.daysAgo == other.daysAgo;
    }
}

File: src/backend/facility/src/main/kotlin/org/taonaw/facility/domain/model/cancellationFee/CancellationFeeRate.kt
package org.taonaw.facility.domain.model.cancellationFee

data class CancellationFeeRate(
        val daysAgo: Int,
        val rate: Double) {

    companion object {
        const val DAYS_AGO_MIN = 0
        const val DAYS_AGO_MAX = 50
        const val RATE_MIN = 0.0
        const val RATE_MAX = 100.0
    }

    init {
        require(daysAgo in DAYS_AGO_MIN..DAYS_AGO_MAX) {
            "何日前かの値は $DAYS_AGO_MIN 以上 $DAYS_AGO_MAX 以下出なければなりません。"
        }
        require(rate in RATE_MIN..RATE_MAX) {
            "レートは $RATE_MIN % 以上 $RATE_MAX % 以下でなければなりません。"
        }
    }

    fun isDuplicated(other: CancellationFeeRate): Boolean {
        return this !== other && daysAgo == other.daysAgo
    }
}